sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/PersistentStorage/StorageHelper.cs

```diff
@@ -12,6 +12,15 @@ internal static class StorageHelper
     {
         private static string? s_defaultStorageDirectory;
 
+        internal static string GetStorageDirectory(string? configuredStorageDirectory, string instrumentationKey)
+        {
+            var storageDirectory = configuredStorageDirectory
+                ?? GetDefaultStorageDirectory()
+                ?? throw new InvalidOperationException("Unable to determine offline storage directory.");
+
+            return Path.Combine(storageDirectory, instrumentationKey);
```

Comment:
I spent some time reviewing this topic.

Guid is 36 characters. 
I haven't found a hash that will significantly reduce the string length (SHA256 is 64 characters, MD5 is 32 characters).
Best I found was using a Base64String which would reduce it to 22 characters.

We could take a substring of the ikey, but then we can't guarantee no collisions. 

--
OpenTelemetry.Extensions.PersistentStorage is creating yet another subdirectory and uses SHA256. ([link](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/b06a6f290fc59a292994e2ba9973eefe0691fa19/src/OpenTelemetry.PersistentStorage.FileSystem/PersistentStorageHelper.cs#L205))

Reply:
For context, here's an example full path (206 characters):

```
"C:\TEMP\00000000-0000-0000-0000-000000000000\a58c82c531c9ee2e54763f35b4c81c1f9b5ebbef091a556cbbf254151475cedd\2023-03-20T222007.2578717Z-0985894a111a41cb8462d7b10bab7cc0.blob@2023-03-20T222107.2645312Z.lock"
```

Reply:
It will be more than 215 characters in most cases.

```
C:\Users\username\AppData\Local\Microsoft\AzureMonitor\3a70e0d7934b71314c58dc3302cb965d0e34883cde72da6d9c279977e8790f1a\2023-03-20T222007.2578717Z-0985894a111a41cb8462d7b10bab7cc0.blob@2023-03-20T222107.2645312Z.lock
```

Reply:
What is the length of the path increased by this change?

Reply:
We can discuss this in our next standup.
In the image below, the gray rows are WITH Ikey.
In all of these examples, we are within the max allowed length.

![image](https://user-images.githubusercontent.com/28785781/227028776-dabce53e-4e02-492a-a1b1-8df026b0df47.png)


However, to have an extra safety net I think we should still reduce the length.
The best I've found is to convert the Guid (36 chars) to a Base64String (22 chars).

```csharp
Convert.ToBase64String(Guid.Parse(inputString).ToByteArray())
```

Reply:
Discussed this in our standup today.
We're going to drive some changes in the PersistentStorage library.
The idea is that we could provide the ikey value to be included in PersistentStorage's hash.

This PR will be on hold while we pursue those changes.

Reply:
The PersistentStorage library has been changed to omit the extra subdirectory.
https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1110

I'll change this PR to handle it's own sub directories.

Reply:
New PersistentStorage library is available and I've updated this PR.